fs/dafsserver/dafsserver.cpp

```diff
@@ -1394,21 +1452,552 @@ class CRemoteDiskReadActivity : public CRemoteDiskBaseActivity
+                    return *curFieldValueFetcher;
+                }
+            };
+            // JCSMORE - it would be good if these were cached/reused (can I assume anything using parent fetcher is single threaded?)
+            return new CIterator(*nested, currentMatch, queryCompoundXPath(fieldNum));
+        }
+        virtual size_t getSize(unsigned fieldNum) const override { throwUnexpected(); }
+        virtual size32_t getRecordSize() const override { throwUnexpected(); }
+    };
+
+    bool checkOpen()
+    {
+        if (!PARENT::checkOpen())
+            return false;
+
+        class CSimpleStream : public CSimpleInterfaceOf<ISimpleReadStream>
+        {
+            Linked<ISerialStream> stream;
+        public:
+            CSimpleStream(ISerialStream *_stream) : stream(_stream)
+            {
+            }
+        // ISimpleReadStream impl.
+            virtual size32_t read(size32_t max_len, void * data) override
+            {
+                size32_t got;
+                const void *res = stream->peek(max_len, got);
+                if (got)
+                {
+                    if (got>max_len)
+                        got = max_len;
+                    memcpy(data, res, got);
+                    stream->skip(got);
+                }
+                return got;
+            }
+        };
+        Owned<ISimpleReadStream> simpleStream = new CSimpleStream(inputStream);
+        if (kind==TAKjsonread)
+            xmlParser.setown(createJSONParse(*simpleStream, xpath, *this, noRoot?ptr_noRoot:ptr_none, useXmlContents));
+        else
+            xmlParser.setown(createXMLParse(*simpleStream, xpath, *this, noRoot?ptr_noRoot:ptr_none, useXmlContents));
+
+        if (!fieldFetcher)
+            fieldFetcher.setown(new CFieldFetcher(*record, nullptr));
+
+        return true;
+    }
+public:
+    IMPLEMENT_IINTERFACE_USING(PARENT);
+
+    CRemoteMarkupReadActivity(IPropertyTree &config, IFileDescriptor *fileDesc, ThorActivityKind _kind) : PARENT(config, fileDesc), kind(_kind)
+    {
+        config.getProp("xpath", xpath);
+        noRoot = config.getPropBool("noRoot");
+    }
+    IColumnProvider *queryMatch() const { return lastMatch; }
+    virtual StringBuffer &getInfoStr(StringBuffer &out) const override
+    {
+        return out.appendf("%s[%s]", getActivityText(kind), fileName.get());
+    }
+// IRemoteReadActivity impl.
+    virtual const void *nextRow(MemoryBufferBuilder &outBuilder, size32_t &retSz) override
+    {
+        if (eofSeen)
+        {
+            retSz = 0;
+            return nullptr;
+        }
+        checkOpen();
+
+        while (xmlParser->next())
+        {
+            if (lastMatch)
+            {
+                ((CFieldFetcher *)fieldFetcher.get())->setCurrentMatch(lastMatch);
+                retSz = translator->translate(outBuilder, *this, *fieldFetcher);
+
+                lastMatch.clear();
+                if (retSz) // could it ever be 0 (skip) ??
```

Comment:
I don't think this can ever be zero.  Probably better to assert that it isn't zero instead.

Reply:
have changed to an assert